### PR TITLE
Adds the most useful component ever: Jammer component

### PR DESCRIPTION
**Pull request by ghost**
I made this to be a nice boy to armhulen. You can now properly stick jammers onto anything. Takes a range and name var.

### CODE REVIEW

File: code/datums/components/jammer.dm [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+/datum/component/jammer/Destroy()
+	if(active)
+		GLOB.active_jammers -= src
+
+/datum/component/jammer/proc/Toggle(mob/user)
+	to_chat(user,"<span class='notice'>You [active ? "deactivate" : "activate"] [jammer_name].</span>")
+	active = !active
+	if(active)
+		GLOB.active_jammers |= src
```
Comment (ninjanomnom) [inline code/datums/components/jammer.dm]: `+=`
This component should handle itself completely, if anyone tries to add it elsewhere they're doing it wrong.

File: code/datums/components/jammer.dm [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/datum/component/jammer
+	var/active = FALSE
+	var/range
+	var/jammer_name
+
+/datum/component/jammer/Initialize(range, var/jammer_name = parent)
+	RegisterSignal(COMSIG_ITEM_ATTACK_SELF, .proc/Toggle)
+	src.range = range
+	src.jammer_name = jammer_name
```
Comment (ninjanomnom) [inline code/datums/components/jammer.dm]: Needs to have a way to start enabled

**Comment by Buggy123** [comment]
But what is it exactly, a radio jammer?

**Comment by Qustinnus** [comment]
a thing that, surprisingly, jams radios in range.

File: code/game/objects/items/devices/radio/radio.dm [hunk 3]
```diff
@@ -240,7 +240,8 @@
 	var/turf/position = get_turf(src)
 
 	var/jammed = FALSE
-	for(var/obj/item/device/jammer/jammer in GLOB.active_jammers)
+	for(var/i in GLOB.active_jammers)
+		var/datum/component/jammer/jammer = i
 		if(get_dist(position,get_turf(jammer)) < jammer.range)
```
Comment (Cruix) [inline code/game/objects/items/devices/radio/radio.dm]: get_turf() on a datum?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/radio/radio.dm (hunk 3 above)
Comment (Qustinnus) [reply code/game/objects/items/devices/radio/radio.dm]: woops, if this doesn't work I can always make it parent.get_turf...

File: code/datums/components/jammer.dm (hunk 1 above)
Comment (Qustinnus) [reply code/datums/components/jammer.dm]: no clue what |= does it was in the original code

File: code/game/objects/items/devices/radio/radio.dm [hunk 4]
```diff
@@ -240,8 +240,9 @@
 	var/turf/position = get_turf(src)
 
 	var/jammed = FALSE
-	for(var/obj/item/device/jammer/jammer in GLOB.active_jammers)
-		if(get_dist(position,get_turf(jammer)) < jammer.range)
+	for(var/i in GLOB.active_jammers)
+		var/datum/component/jammer/jammer = i
+		if(get_dist(position,get_turf(jammer.parent)) < jammer.range)
```
Comment (Cruix) [inline code/game/objects/items/devices/radio/radio.dm]: What does get_dist() return if one of the objects is null or in nullspace?

Comment (Qustinnus) [reply code/game/objects/items/devices/radio/radio.dm]: no clue but it caps at 127 so nullspace should be fine.

Comment (Cruix) [reply code/game/objects/items/devices/radio/radio.dm]: But if it returns 0 then a jammer in nullspace will jam every radio everywhere

Comment (Cruix) [reply code/game/objects/items/devices/radio/radio.dm]: Does this check for z-level anywhere?

Comment (Qustinnus) [reply code/game/objects/items/devices/radio/radio.dm]: that would be fairly comedical and I'm not sure how to fix that. I guess fix it by not having things go to nullspace :^)

Comment (Qustinnus) [reply code/game/objects/items/devices/radio/radio.dm]: also get_dist is a byond proc I have no clue if it checks for z-level

Comment (Cruix) [reply code/game/objects/items/devices/radio/radio.dm]: I know from experience that it does not, but does the rest of the code here?

File: code/datums/components/jammer.dm [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/datum/component/jammer
+	var/active = FALSE
+	var/range
+	var/jammer_name
+
+/datum/component/jammer/Initialize(range, var/jammer_name = parent)
+	if(istype(parent, obj/item))
+		RegisterSignal(COMSIG_ITEM_ATTACK_SELF, .proc/Toggle)
+	src.range = range
```
Comment (kevinz000) [inline code/datums/components/jammer.dm]: >>>SRC

Comment (Qustinnus) [reply code/datums/components/jammer.dm]: range = range

OK kevinz

File: code/game/objects/items/devices/traitordevices.dm [hunk 6]
```diff
@@ -237,14 +237,12 @@ effective or pretty fucking useless.
 	name = "radio jammer"
 	desc = "Device used to disrupt nearby radio communication."
 	icon_state = "jammer"
-	var/active = FALSE
-	var/range = 12
+
+/obj/item/device/jammer/Initialize()
+	AddComponent(/datum/component/jammer, 12, src)
 
 /obj/item/device/jammer/attack_self(mob/user)
```
Comment (Qustinnus) [inline code/game/objects/items/devices/traitordevices.dm]: Lol woops i forgot i need to remove this, ill do that tommorow because im sleeping now zz

**Review by ghost — COMMENTED** [review]
im dumb

File: code/game/objects/items/devices/radio/radio.dm (hunk 4 above)
Comment (AnturK) [reply code/game/objects/items/devices/radio/radio.dm]: Oh whoops forgot to check Z here. Yeah add z check please.

**Comment by Buggy123** [comment]
CBA to read the code, is this a AOE jam, a jam of a given target, or what?

Actually, now that i think about it that's a neat idea. Give us the option to jam every headset in a AOE while using lots of power, or just jam one specified target within range.

**Comment by evsey9** [comment]
shouldnt this be a traitor item

**Comment by Arianya** [comment]
Should make raspberry jam leak out of affected radios

**Comment by Arianya** [comment]
Not just because its funny but also because it gives feedback as to whats causing the jamming

File: code/datums/components/jammer.dm (hunk 5 above)
Comment (Supermichael777) [reply code/datums/components/jammer.dm]: this will set jammer's range var to... jammer's range var

10/10

Comment (Supermichael777) [reply code/datums/components/jammer.dm]: src.a is the same as a whats the point of this

Comment (JJRcop) [reply code/datums/components/jammer.dm]: The arg is called range, it should be called something else.

**Comment by praisenarsie** [comment]
@evsey9 it is already, its just refactor of jammer code

File: code/datums/components/jammer.dm (hunk 5 above)
Comment (Qustinnus) [reply code/datums/components/jammer.dm]: hnng I remember seeing this somewhere, alright I'll fix that

**Comment by Qustinnus** [comment]
^ what he said.

this doesn't add anything, it just makes what we have easier to use

File: code/datums/components/jammer.dm (hunk 5 above)
Comment (ninjanomnom) [reply code/datums/components/jammer.dm]: @Supermichael777 not in all circumstances, range is an arg in this var which means plain `range` will use that first. `src.` tells the code that we want the arg from the jammer instead.

Comment (Qustinnus) [reply code/datums/components/jammer.dm]: wait, I wasn't wrong for once?

Comment (ninjanomnom) [reply code/datums/components/jammer.dm]: No but @JJRcop is right that you should rename the arg

**Comment by Qustinnus** [comment]
@jughu what are u on about :x

**Comment by DoctoreMad** [comment]
@jughu This PR doesn't do what you think it does

File: code/datums/components/jammer.dm [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/datum/component/jammer
+	var/active = FALSE
+	var/range
+	var/name
+
+/datum/component/jammer/Initialize(jammer_range, var/jammer_name = parent)
+	if(istype(parent, /obj/item))
+		RegisterSignal(COMSIG_ITEM_ATTACK_SELF, .proc/Toggle)
+	range = jammer_range
+	name = jammer_name
```
Comment (ninjanomnom) [inline code/datums/components/jammer.dm]: You need to add an arg for starting active for things that don't have a toggle.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
Weren't you doing this pr partialy for putting it on mobs @Qustinnus ?

**Comment by Qustinnus** [comment]
yes for armhulen, why? did I do something wrong?

File: code/datums/components/jammer.dm [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/datum/component/jammer
+	var/active = FALSE
+	var/range
+	var/name
+
+/datum/component/jammer/Initialize(jammer_range, jammer_name = parent, start_activated = FALSE)
```
Comment (ninjanomnom) [inline code/datums/components/jammer.dm]: `jammer_name = "[parent]"`

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
You'll need to have the component give an action button to to the mob it's applied to.

**Comment by Qustinnus** [comment]
what if the mob has passive jamming though..?

**Comment by ninjanomnom** [comment]
Then don't give the action button? It'd just be a simple arg

**Comment by Armhulen** [comment]
go floyd go disco infernoooo

also what if i want the buttons to have a special background and all that but if it makes life easier lets do it

**Comment by AnturK** [comment]
This action button coming or not ?

**Comment by Qustinnus** [comment]
I'll finish this once I resolve some business IRL

**Pull request closed without merging**